geo: initial framework to parse WKT from GEOS

otan opened the pull request:
This PR introduces the ability to parse WKT using the GEOS library. It
is the only library I can tell that can decode WKT, so everything will
have to go through that (for now). We're planning to store everything in
EWKB, and everything can be parsed off WKB which is why I've made
that the base type for now.

In future, the geos package will take in WKBs in a new package `geomfn`,
and transform into GEOSGeometry for operations without ever returning
*GEOSGeometry back to the user. This is necessary as we have no control
on the lifetime of a Geometry object.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46867)
<!-- Reviewable:end -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/geo/geos/geos_unix.h, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1X4xB2gQLK4j_WnX:-M3r1X4xB2gQLK4j_WnY:b1my3xk) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.h#L18)):*
> ```objc
> 
> // Data Types adapted from `capi/geos_c.h.in` in GEOS.
> typedef void* GEOSHandle;
> ```

should i be putting `_t` at the end of these? 
___
*[pkg/geo/geos/geos_unix.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1HsZ5VZm8dZaGTdG:-M3r1HsZ5VZm8dZaGTdH:b-aev19g) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.go#L56)):*
> ```Go
> 	}
> 	for _, extraLoc := range []string{
> 		"../../../lib/libgeos_c.dylib",
> ```

is there a better way to do this?
___
*[pkg/geo/geos/geos_unix_test.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1RwBArNJWIa6ggDg:-M3r1RwBArNJWIa6ggDh:bglhvkd) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix_test.go#L32)):*
> ```Go
> 		holder := bootstrap(defaultGEOSLocations)
> 		assert.True(t, holder.valid)
> 		fmt.Printf("%#v\n", defaultGEOSLocations)
> ```

i will delete this.
___
*[pkg/geo/geos/geos_windows.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1K952xlIpzXwIxMP:-M3r1K952xlIpzXwIxMQ:b-3qdojq) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_windows.go#L20)):*
> ```Go
> )
> 
> func WKTToWKB(wkt geo.WKT) (geo.WKB, error) {
> ```

any way to avoid doing this annoying thing for each function we define?


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
Generally looks sane. I need to take a look at the circular dependency you mentioned. This is a first pass of comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-:-M3r68oq-2teeiTsW5EP:bqx7grp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geomfn/geomfn_test.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r3WSf96rJZx28nycz:-M3r3WSf96rJZx28nyd-:b-5oxa1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geomfn/geomfn_test.go#L29)):*
> ```Go
> 		{
> 			"POINT(1.0 1.0)",
> 			geo.NewGeometry(geo.EWKB([]byte{0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xf0, 0x3f, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xf0, 0x3f})),
> ```

A string constant would be slightly more compact: `[]byte("\x01\x01\x00\x00\x00...")`
___
*[pkg/geo/geos/geos_unix.h, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1X4xB2gQLK4j_WnX:-M3r24tqEnXnaLcypXo_:b-loel6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.h#L18)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

should i be putting `_t` at the end of these? 
</blockquote></details>

Not in my opinion.
___
*[pkg/geo/geos/geos_unix.h, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r2iYu2zl-11OhD0C3:-M3r2iYu2zl-11OhD0C4:bff5pqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.h#L53)):*
> ```objc
>   GEOSWKBWriter_destroy_r GEOSWKBWriter_destroy_r;
>   GEOSWKBWriter_write_r GEOSWKBWriter_write_r;
> } GEOSLib;
> ```

I don't think you need to expose this structure here as we'll never be accessing the internals from Go. Instead, you do something like:

```
typedef struct GEOSLib GEOSLib;
```

And then in some other private header file (not accessed by Go) you can define `GEOSLib`:

```
struct GEOSLib {
  ...
};
```
___
*[pkg/geo/geos/geos_unix.h, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r27123nruLj6B2hIe:-M3r27123nruLj6B2hIf:b-tcshwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.h#L57)):*
> ```objc
> // bootstrap initializes the provided GEOSLib with GEOS using dlopen/dlsym.
> // Returns a string containing an error if an error was found.
> char* bootstrap(GEOSLib *geos_lib, char *lib_location);
> ```

C has a single global namespace, which is why C libraries often prefix function names with a constant string. For `libroach`, we use `DB`. I would have suggested using `GEOS` here, but that is too likely to collide with the libgeos functions (yes, I think we have to worry about collision even for dynamically loaded libs).

We have been using CamelCase style for function names in other C/C++ code. I think we should do similar here for consistency.
___
*[pkg/geo/geos/geos_unix.go, line 56 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1HsZ5VZm8dZaGTdG:-M3r3yll-rJAZsuuUXYl:b8k0bye) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.go#L56)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

is there a better way to do this?
</blockquote></details>

I think you should define a var for the shared library extension:

```
dynLibExt := "so"
if runtime.GOOS == "Darwin" {
  dynLibExt = "dylib"
}
```
___
*[pkg/geo/geos/geos_unix.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r50_r4M8KIw5eNkgW:-M3r50_r4M8KIw5eNkgX:b-as7zvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.go#L88)):*
> ```Go
> 	cWKT := C.CString(string(wkt))
> 	defer C.free(unsafe.Pointer(cWKT))
> 	cWKB := C.wkt_to_wkb(&holder.geosLib, cWKT, &dataLen)
> ```

Our experience with RocksDB is that it is worthwhile minimizing the Cgo calls. For instance, we could pass the `wkt` as-is to C and do the copy and append of the NUL byte to create a C string inside `C.wkt_to_wkb`. How do you pass a Go string to C function? You can do something similar to `pkg/storage.goToCSlice`. That routine doesn't do an allocation, but passes the string buffer and length. 
___
*[pkg/geo/geos/geos_unix.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r5YnlEhwTRy-OWkIH:-M3r5Ynm8OTKNhesPDcm:bszhsk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.go#L89)):*
> ```Go
> 	defer C.free(unsafe.Pointer(cWKT))
> 	cWKB := C.wkt_to_wkb(&holder.geosLib, cWKT, &dataLen)
> 	if unsafe.Pointer(cWKB) == C.NULL {
> ```

Can this be `cWKB == nil` instead? Or does something prevent that from working?


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
couple nits: i didn't look at any of the cgo stuff

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-:-M3rAv5g3yyQsRUKbB_5:b61gm55)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geo.go, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3rA5yWDsP1pCgeFYYT:-M3rA5yWDsP1pCgeFYYU:blypfpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/b75d35bbc5ba8d11132c06c08572c5f41d2da9d1/pkg/geo/geo.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> // Package geo is a package that contains the base types for
> ```

nit: slightly redundant. should be fine to s/is a package that//
___
*[pkg/geo/geo.go, line 38 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3rADUk49JzHzlTb1ZR:-M3rADUk49JzHzlTb1ZS:b-us3dc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b75d35bbc5ba8d11132c06c08572c5f41d2da9d1/pkg/geo/geo.go#L38)):*
> ```Go
> 
> // EWKB is the Extended Well Known Bytes form of a spatial object.
> // It is safe to cast any WKB to any EWKB, but EWKB to WKB.
> ```

nit: do you mean that EWKB to WKB is _not_ safe?

Also I'd just make a `func WKBToEWKB(wkb WKB) EWKB)` instead of this comment
___
*[pkg/geo/geos/geos_windows.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1K952xlIpzXwIxMP:-M3rApin186fpUTDjtjJ:bluikip) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_windows.go#L20)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

any way to avoid doing this annoying thing for each function we define?
</blockquote></details>

codegen?
___
*[pkg/geo/geos/geos_windows.go, line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3rA_xYBaPowSZ9loq8:-M3rA_xYBaPowSZ9loq9:bqr9olu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b75d35bbc5ba8d11132c06c08572c5f41d2da9d1/pkg/geo/geos/geos_windows.go#L21)):*
> ```Go
> 
> func WKTToWKB(wkt geo.WKT) (geo.WKB, error) {
> 	return nil, errors.Newf("not available in windows")
> ```

make this an `errors.UnimplementedError` with a pointer to a placeholder issue for windows geospatial support


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @petermattis)

___
*[pkg/geo/geomfn/geomfn_test.go, line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r3WSf96rJZx28nycz:-M3rIc5uB3vbZ_g0UKvD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geomfn/geomfn_test.go#L29)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

A string constant would be slightly more compact: `[]byte("\x01\x01\x00\x00\x00...")`
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.h, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r2iYu2zl-11OhD0C3:-M3rKJQ76UedccSOJpA5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.h#L53)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I don't think you need to expose this structure here as we'll never be accessing the internals from Go. Instead, you do something like:

```
typedef struct GEOSLib GEOSLib;
```

And then in some other private header file (not accessed by Go) you can define `GEOSLib`:

```
struct GEOSLib {
  ...
};
```


</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.h, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r27123nruLj6B2hIe:-M3rOJP575UAhuGbiQv9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.h#L57)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

C has a single global namespace, which is why C libraries often prefix function names with a constant string. For `libroach`, we use `DB`. I would have suggested using `GEOS` here, but that is too likely to collide with the libgeos functions (yes, I think we have to worry about collision even for dynamically loaded libs).

We have been using CamelCase style for function names in other C/C++ code. I think we should do similar here for consistency.
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r50_r4M8KIw5eNkgW:-M3rkO7l4OEQ6-XrR05E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.go#L88)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Our experience with RocksDB is that it is worthwhile minimizing the Cgo calls. For instance, we could pass the `wkt` as-is to C and do the copy and append of the NUL byte to create a C string inside `C.wkt_to_wkb`. How do you pass a Go string to C function? You can do something similar to `pkg/storage.goToCSlice`. That routine doesn't do an allocation, but passes the string buffer and length. 
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r5YnlEhwTRy-OWkIH:-M3rP0DM0CtP5gwDoA_b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.go#L89)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can this be `cWKB == nil` instead? Or does something prevent that from working?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> Reviewed 9 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @petermattis)

___
*[pkg/geo/geopb/types.go, line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3sO7iiD8jv3gJu6xcM:-M3sO7ijDoTI811aHMVG:bxhyl5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/31828870f41919d8f5dee78cbd53c8e377daca88/pkg/geo/geopb/types.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> package geopb
> ```

Is this package going to contain protos?
___
*[pkg/geo/geos/geos_unix.go, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3sMeE-8tYAdenhzWl9:-M3sMeE-8tYAdenhzWlA:b28mf2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/31828870f41919d8f5dee78cbd53c8e377daca88/pkg/geo/geos/geos_unix.go#L116)):*
> ```Go
> 	}
> 	wkb := cSliceToUnsafeGoBytes(cWKB)
> 	defer C.free(unsafe.Pointer(cWKB.data))
> ```

isn't this going to free the memory that the slice is pointing at?
I think `rocksdb.go` does such freeing after making a copy.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @petermattis, and @sumeerbhola)

___
*[pkg/geo/geopb/types.go, line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3sO7iiD8jv3gJu6xcM:-M3sRCPO2ajeGLVUmzEZ:bhequ85) ([raw file](https://github.com/cockroachdb/cockroach/blob/31828870f41919d8f5dee78cbd53c8e377daca88/pkg/geo/geopb/types.go#L11)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is this package going to contain protos?
</blockquote></details>

yes, eventually.
___
*[pkg/geo/geos/geos_unix.go, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3sMeE-8tYAdenhzWl9:-M3sRE1S9CzQW06qdR3_:bvs5ek) ([raw file](https://github.com/cockroachdb/cockroach/blob/31828870f41919d8f5dee78cbd53c8e377daca88/pkg/geo/geos/geos_unix.go#L116)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

isn't this going to free the memory that the slice is pointing at?
I think `rocksdb.go` does such freeing after making a copy.
</blockquote></details>

yeah explains the test failure. thanks for pointing it out.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @petermattis, and @sumeerbhola)

___
*[pkg/geo/geo_test.go, line 50 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3v-g5DC0ypTWkMR5cT:-M3v-g5E6cATekIKNCFE:bmauvjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geo_test.go#L50)):*
> ```Go
> 			} else {
> 				assert.NoError(t, err)
> 				assert.Equal(t, tc.expected, g)
> ```

I've been using `testify/require` in my tests, rather than `testify/assert`. TIL the difference is that `assert.*` is equivalent to `t.Error*`, while `require.*` is equivalent to `t.Fatal*`. I mention this because I'm not sure it is worthwhile to perform this second check if the first one failed. 
___
*[pkg/geo/geos/geos_unix.h, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r2iYu2zl-11OhD0C3:-M3uuwZuBtC_WLOx68wm:bsv8usl) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.h#L53)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The structure looks good. I'm going to nit on the `CockroachGEOS` prefix. That is going to get tiresome to type (and read) soon. Perhaps we can have a convention like `CR_<libraryname>` for those handles. So this would be `CR_GEOS` (no need for the lib suffix). Just brainstorming here. Open to other ideas as well.
___
*[pkg/geo/geos/geos_unix.c, line 55 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uwIZJ4kRgMIl71vnR:-M3uwIZJ4kRgMIl71vnS:baa7yby) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.c#L55)):*
> ```C
> };
> 
> char *CockroachGEOSStringToCString(CockroachGEOSString goStr) {
> ```

I believe our C-style associates the pointer with the type. That is, `char* CockroachGEOSStringToCString`. Put rather than argue about this, we already have a formatter defined. See `make c-deps-fmt`. We should make sure this code adheres to whatever that formatter produces. Right now it only formats code in `c-deps/libroach`. That is something else to fix (not necessarily for this PR).
___
*[pkg/geo/geos/geos_unix.c, line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uvi0iDuHw8fIAfi7E:-M3uvi0iDuHw8fIAfi7F:b-dysbli) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.c#L62)):*
> ```C
> }
> 
> CockroachGEOSLib *CockroachGEOSInitLib(CockroachGEOSString loc) {
> ```

I think we'll want to return an error from this function. We can do that be passing a pointer to a pointer for the return value of  `CockroachGEOSLib`:

```
char* CockroachGEOSInitLib(CockroachGEOSString loc, CockroachGEOSLib** result) {
  ...
}
```
___
*[pkg/geo/geos/geos_unix.c, line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uvN6a3aBBs28kEgHt:-M3uvN6b1dIzzxBiFTPs:b-peadh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.c#L78)):*
> ```C
>   lib->GEOS_init_r = (GEOS_init_r) dlsym(dlHandle, "GEOS_init_r");
>   if ((error = dlerror()) != NULL) {
>     free(lib);
> ```

I suspect we may want to rename this file to `geos_unix.cc` so we can use C++. Specifically, I would store `lib` in a `unique_ptr` and then detach it on successful return. This would avoid having to do this `free(lib)` on every return path. It would also allow us to get rid of the use of manual memory management for C strings.
___
*[pkg/geo/geos/geos_unix.go, line 47 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxtgn3kkTssGy0RT7:-M3uxtgn3kkTssGy0RT8:b-9id4qs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L47)):*
> ```Go
> var defaultGEOSLocations []string
> 
> var globalCockroachGEOSLib *C.CockroachGEOSLib
> ```

Nit: we don't use the term "global" to refer to globals anywhere else that I'm aware of.
___
*[pkg/geo/geos/geos_unix.go, line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxGanEi4TkRMaMUNu:-M3uxGanEi4TkRMaMUNv:b-enyq4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L57)):*
> ```Go
> 	split := strings.Split(cwd, "/")
> 	for i := 0; i < len(split); i++ {
> 		nextPath := "/" + path.Join(split[:i]...)
> ```

I suspect you want to use the functions in `filepath` for these path manipulations. Specifically, `filepath.Dir` will get you the parent directory.
___
*[pkg/geo/geos/geos_unix.go, line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxcoG1f_Y4dH1e_XW:-M3uxcoG1f_Y4dH1e_XX:b95y5ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L59)):*
> ```Go
> 		nextPath := "/" + path.Join(split[:i]...)
> 		if runtime.GOOS == "darwin" {
> 			nextPath = path.Join(nextPath, "lib", "libgeos_c.dylib")
> ```

`s/path/filepath/g`. The "path" package works on slash separated paths. The "filepath" package uses the OS path separator. Small difference, but "filepath" is what you want if we're manipulating paths for the current OS.
___
*[pkg/geo/geos/geos_windows.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1K952xlIpzXwIxMP:-M3uyXzO5P9n5s84inBQ:b5d9tqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_windows.go#L20)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

codegen?
</blockquote></details>

Yes, I would look into codegen. Finding the exported functions in `geos_unix.go` shouldn't be difficult (can be done with grep). Determining the return arguments is a bit more irritating. I don't have an immediate thought there, but I'm sure something can be done.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geos/geos_windows.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1K952xlIpzXwIxMP:-M3vHv8S0WJXMeIZxqNp:b8vcqlm) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_windows.go#L20)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yes, I would look into codegen. Finding the exported functions in `geos_unix.go` shouldn't be difficult (can be done with grep). Determining the return arguments is a bit more irritating. I don't have an immediate thought there, but I'm sure something can be done.
</blockquote></details>

ok, i'll save this for a future effort (or just make it work with windows...)


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geos/geos_unix.go, line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxGanEi4TkRMaMUNu:-M3vKgT78dx17fTGOE-i:bgfciio) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L57)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I suspect you want to use the functions in `filepath` for these path manipulations. Specifically, `filepath.Dir` will get you the parent directory.
</blockquote></details>

weird, it didnt work for me yesterday but i may have had a day of strange bugs. fixed!


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.go, line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxGanEi4TkRMaMUNu:-M3vLFFi4hO5hT75Z8cl:b668goe) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L57)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

weird, it didnt work for me yesterday but i may have had a day of strange bugs. fixed!
</blockquote></details>

Did you forget to push? I'm not seeing any changes here.
___
*[pkg/geo/geos/geos_windows.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r1K952xlIpzXwIxMP:-M3vLIWl6frjJTci3dFI:b-9chsef) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_windows.go#L20)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

ok, i'll save this for a future effort (or just make it work with windows...)
</blockquote></details>

Ack. Fine to save this for a future PR.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @sumeerbhola)

___
*[pkg/geo/geo_test.go, line 50 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3v-g5DC0ypTWkMR5cT:-M3vGMuw1qwUxlsw-lLU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geo_test.go#L50)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I've been using `testify/require` in my tests, rather than `testify/assert`. TIL the difference is that `assert.*` is equivalent to `t.Error*`, while `require.*` is equivalent to `t.Fatal*`. I mention this because I'm not sure it is worthwhile to perform this second check if the first one failed. 
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.h, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3r2iYu2zl-11OhD0C3:-M3vH2hvEs3VqodRL_-u:b-s72kwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb95c11606001e4af8ddc6f58aecff926f28c4b3/pkg/geo/geos/geos_unix.h#L53)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The structure looks good. I'm going to nit on the `CockroachGEOS` prefix. That is going to get tiresome to type (and read) soon. Perhaps we can have a convention like `CR_<libraryname>` for those handles. So this would be `CR_GEOS` (no need for the lib suffix). Just brainstorming here. Open to other ideas as well.
</blockquote></details>

i'm going to do CR_GEOS_ - i think that's reasonable.
___
*[pkg/geo/geos/geos_unix.c, line 55 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uwIZJ4kRgMIl71vnR:-M3vRsVz7fHVlDvFWwgt:b-oq4jl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.c#L55)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I believe our C-style associates the pointer with the type. That is, `char* CockroachGEOSStringToCString`. Put rather than argue about this, we already have a formatter defined. See `make c-deps-fmt`. We should make sure this code adheres to whatever that formatter produces. Right now it only formats code in `c-deps/libroach`. That is something else to fix (not necessarily for this PR).
</blockquote></details>

i've formatted it for now. 
it seems to prefer `char *CR_GEOS_StringToCString(`.
(also seems like it hasn't run on libroach in a while :3)
___
*[pkg/geo/geos/geos_unix.c, line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uvi0iDuHw8fIAfi7E:-M3vXT0x0KECmz_8r3P4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.c#L62)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think we'll want to return an error from this function. We can do that be passing a pointer to a pointer for the return value of  `CockroachGEOSLib`:

```
char* CockroachGEOSInitLib(CockroachGEOSString loc, CockroachGEOSLib** result) {
  ...
}
```
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.c, line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uvN6a3aBBs28kEgHt:-M3vL94XBtOLJxyGs8Rt:buebwfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.c#L78)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I suspect we may want to rename this file to `geos_unix.cc` so we can use C++. Specifically, I would store `lib` in a `unique_ptr` and then detach it on successful return. This would avoid having to do this `free(lib)` on every return path. It would also allow us to get rid of the use of manual memory management for C strings.
</blockquote></details>

i've gotten 

```
Undefined symbols for architecture x86_64:
  "_CR_GEOS_Init", referenced from:
      __cgo_746a18938317_Cfunc_CR_GEOS_Init in _x002.o
     (maybe you meant: __cgo_746a18938317_Cfunc_CR_GEOS_Init)
  "_CR_GEOS_WKTToWKB", referenced from:
      __cgo_746a18938317_Cfunc_CR_GEOS_WKTToWKB in _x002.o
     (maybe you meant: __cgo_746a18938317_Cfunc_CR_GEOS_WKTToWKB)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
FAIL	github.com/cockroachdb/cockroach/pkg/geo/geos [build failed]
```

when trying to do this.
___
*[pkg/geo/geos/geos_unix.go, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3sMeE-8tYAdenhzWl9:-M3vHtspBHVqv4epnrni:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31828870f41919d8f5dee78cbd53c8e377daca88/pkg/geo/geos/geos_unix.go#L116)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

yeah explains the test failure. thanks for pointing it out.
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.go, line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxGanEi4TkRMaMUNu:-M3vKq7g1Xn5qNkqVZ5y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L57)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

weird, it didnt work for me yesterday but i may have had a day of strange bugs. fixed!
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.go, line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxcoG1f_Y4dH1e_XW:-M3vKkx5EqapZwRc_Y-7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L59)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/path/filepath/g`. The "path" package works on slash separated paths. The "filepath" package uses the OS path separator. Small difference, but "filepath" is what you want if we're manipulating paths for the current OS.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.c, line 78 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uvN6a3aBBs28kEgHt:-M3vvyrn7jYbx8NKILmQ:bs932mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.c#L78)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i've gotten 

```
Undefined symbols for architecture x86_64:
  "_CR_GEOS_Init", referenced from:
      __cgo_746a18938317_Cfunc_CR_GEOS_Init in _x002.o
     (maybe you meant: __cgo_746a18938317_Cfunc_CR_GEOS_Init)
  "_CR_GEOS_WKTToWKB", referenced from:
      __cgo_746a18938317_Cfunc_CR_GEOS_WKTToWKB in _x002.o
     (maybe you meant: __cgo_746a18938317_Cfunc_CR_GEOS_WKTToWKB)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
FAIL	github.com/cockroachdb/cockroach/pkg/geo/geos [build failed]
```

when trying to do this.
</blockquote></details>

Fixed - thanks for the extern pointer.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
A few C++ gotchas to address. I think this is nearly good to go.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-:-M3w0OOa89O6jTcS--2d:b-m0i3qf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @petermattis, and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.cc, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3vyzGh2gGFyeQeaCxT:-M3vyzGh2gGFyeQeaCxU:bm07xst) ([raw file](https://github.com/cockroachdb/cockroach/blob/2193be696cac9a9462bca7f4fe2a4b4670406583/pkg/geo/geos/geos_unix.cc#L67)):*
> ```cpp
>   char *error;
> 
>   const char *locStr = CR_GEOS_StringToString(loc).c_str();
> ```

I'm pretty sure this is no bueno. The lifetime of the C++ string object ends when the statement ends. The return value from `.c_str()` cannot live past that lifetime. Instead, you should assign the C++ string to a variable:

```
auto locStr = CR_GEO_StringToString(loc);
void *dlHandle = dlopen(locStr.c_str(), RTLD_LAZY);
```

It is possible that the lifetime of the temporary string exists for longer than I expect, but I'd rather not rely on the amount of subtlety. 
___
*[pkg/geo/geos/geos_unix.cc, line 70 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3vzInREsMw1FCyL8vx:-M3vzInREsMw1FCyL8vy:bqt3pce) ([raw file](https://github.com/cockroachdb/cockroach/blob/2193be696cac9a9462bca7f4fe2a4b4670406583/pkg/geo/geos/geos_unix.cc#L70)):*
> ```cpp
>   void *dlHandle = dlopen(locStr, RTLD_LAZY);
>   if (!dlHandle) {
>     return (char *)std::string("loc does not exist").c_str();
> ```

This is also no bueno for a similar reason to the above comment. You're creating a `std::string` temporary, then extracting a C string which is a pointer into the `std::string` buffer. The return value from `.c_str()` will be invalid (the underlying memory freed) when the function returns.
___
*[pkg/geo/geos/geos_unix.cc, line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3vziCWDdvya51t4WxA:-M3vziCWDdvya51t4WxB:b-5ztow7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2193be696cac9a9462bca7f4fe2a4b4670406583/pkg/geo/geos/geos_unix.cc#L79)):*
> ```cpp
>   ret->CR_GEOS_init_r = (CR_GEOS_init_r)dlsym(dlHandle, "GEOS_init_r");
>   if ((error = dlerror()) != NULL) {
>     return error;
> ```

Is the caller expected to free the returned error? I think the answer is no here. We need to make that explicit in the contract for this function. That implies to me that the failure to `dlopen` can do: `return "loc does not exist"`.  Or perhaps: `return "dlopen failed"` (since perhaps `loc` does exist, but it couldn't be loaded for some reason).

Should we be `dlclose` on these error paths? Something we could do there is to add a destructor to the `CR_GEOS` struct.
___
*[pkg/geo/geos/geos_unix.go, line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxcoG1f_Y4dH1e_XW:-M3w-Crg4XNIL8yWMDLw:b5x7swk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L59)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

There are still two calls to `path.Join` here which should be `filepath.Join`.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.cc, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3vyzGh2gGFyeQeaCxT:-M3w2sg58SNhLMzpq9iG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2193be696cac9a9462bca7f4fe2a4b4670406583/pkg/geo/geos/geos_unix.cc#L67)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm pretty sure this is no bueno. The lifetime of the C++ string object ends when the statement ends. The return value from `.c_str()` cannot live past that lifetime. Instead, you should assign the C++ string to a variable:

```
auto locStr = CR_GEO_StringToString(loc);
void *dlHandle = dlopen(locStr.c_str(), RTLD_LAZY);
```

It is possible that the lifetime of the temporary string exists for longer than I expect, but I'd rather not rely on the amount of subtlety. 
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.cc, line 70 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3vzInREsMw1FCyL8vx:-M3w2qEK2V2oQe_yaI_L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2193be696cac9a9462bca7f4fe2a4b4670406583/pkg/geo/geos/geos_unix.cc#L70)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is also no bueno for a similar reason to the above comment. You're creating a `std::string` temporary, then extracting a C string which is a pointer into the `std::string` buffer. The return value from `.c_str()` will be invalid (the underlying memory freed) when the function returns.
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.cc, line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3vziCWDdvya51t4WxA:-M3w2qxC2GWWMYOnRXSy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2193be696cac9a9462bca7f4fe2a4b4670406583/pkg/geo/geos/geos_unix.cc#L79)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is the caller expected to free the returned error? I think the answer is no here. We need to make that explicit in the contract for this function. That implies to me that the failure to `dlopen` can do: `return "loc does not exist"`.  Or perhaps: `return "dlopen failed"` (since perhaps `loc` does exist, but it couldn't be loaded for some reason).

Should we be `dlclose` on these error paths? Something we could do there is to add a destructor to the `CR_GEOS` struct.
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.go, line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uxcoG1f_Y4dH1e_XW:-M3w1Bdh97ul8Z486kj9:b-lc5t1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.go#L59)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There are still two calls to `path.Join` here which should be `filepath.Join`.
</blockquote></details>

woops! got them.


<!-- Sent from Reviewable.io -->

otan committed:
geo: initial framework to parse WKT from GEOS

This PR introduces the ability to parse WKT using the GEOS library. It
is the only library I can tell that can decode WKT, so everything will
have to go through that (for now). We're planning to store everything in
EWKB, and everything can be parsed off WKB which is why I've made
that the base type for now.

In future, the geos package will take in WKBs in a new package `geomfn`,
and transform into GEOSGeometry for operations without ever returning
*GEOSGeometry back to the user. This is necessary as we have no control
on the lifetime of a Geometry object.

Release note: None

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @petermattis, and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.c, line 55 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3uwIZJ4kRgMIl71vnR:-M3w4sUN9e_cHTPrYgh1:bjds3yz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5bbfb760f4d357140a695e508e89dbd6cb646e6/pkg/geo/geos/geos_unix.c#L55)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i've formatted it for now. 
it seems to prefer `char *CR_GEOS_StringToCString(`.
(also seems like it hasn't run on libroach in a while :3)
</blockquote></details>

The automatic formatter is the boss.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> module nits and figuring out what is going on with CI.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-:-M3w5_n8DYDovcs8qjIl:bbwwimg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.cc, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3vyzGh2gGFyeQeaCxT:-M3w5AmO8YIeReWUqF-h:bb054wo) ([raw file](https://github.com/cockroachdb/cockroach/blob/2193be696cac9a9462bca7f4fe2a4b4670406583/pkg/geo/geos/geos_unix.cc#L67)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Did `auto locStr` not work? The `auto` stuff might be from a newer C++ version. Should work. We use that in `libroach`. Saves needing to know the type.
___
*[pkg/geo/geos/geos_unix.cc, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3w5NJK7qFEo6AXpiOO:-M3w5NJLAZx5aEZnA4FB:bmnjkdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb4039b9d9515c727f28954900a66ebf38f5064c/pkg/geo/geos/geos_unix.cc#L60)):*
> ```cpp
> 
>   ~CR_GEOS() {
>     if (this->dlHandle != NULL) {
> ```

Nit: I'd be consistent here and always omit the `this->` when referring to `dlHandle`.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geos/geos_unix.cc, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3w5NJK7qFEo6AXpiOO:-M3w9ATgDTpIfeHI9LYZ:bymxdey) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb4039b9d9515c727f28954900a66ebf38f5064c/pkg/geo/geos/geos_unix.cc#L60)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: I'd be consistent here and always omit the `this->` when referring to `dlHandle`.
</blockquote></details>

didn't realise we don't use this-> by default. is there any vague style guide we have for cpp at CRDB?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @petermattis and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.cc, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3vyzGh2gGFyeQeaCxT:-M3w94eD5mz8Ilge3yYZ:bnl50is) ([raw file](https://github.com/cockroachdb/cockroach/blob/2193be696cac9a9462bca7f4fe2a4b4670406583/pkg/geo/geos/geos_unix.cc#L67)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did `auto locStr` not work? The `auto` stuff might be from a newer C++ version. Should work. We use that in `libroach`. Saves needing to know the type.
</blockquote></details>

it does after the CXX flags got put in.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan, @petermattis, and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.h, line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3w9_zb9pHxBDrVEPTM:-M3w9_zb9pHxBDrVEPTN:b4e87c0) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a3054d68b2a4f7109627c744f23b2fa45b15bc/pkg/geo/geos/geos_unix.h#L42)):*
> ```objc
> 
> // CR_GEOS_WKTToWKB converts a given WKT into it's WKB form.
> CR_GEOS_Slice CR_GEOS_WKTToWKB(CR_GEOS *lib, CR_GEOS_String wkt);
> ```

In `libroach` the difference between `DBSlice` and `DBString` is ownership. A returned `DBSlice` implies the underlying memory does not need to be freed. A returned `DBString` must have `DBString::data` freed by the caller of the function. I think it would be useful to maintain that distinction here. So `s/CR_GEOS_Slice/CR_GEOS_String/g`.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan, @petermattis, and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.go, line 115 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3wAEEJA843fV-D3qnN:-M3wAEEJA843fV-D3qnO:b-s0g0t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a3054d68b2a4f7109627c744f23b2fa45b15bc/pkg/geo/geos/geos_unix.go#L115)):*
> ```Go
> 
> // WKTToWKB parses a WKT into WKB using the GEOS library.
> func WKTToWKB(wkt geopb.WKT) (geopb.WKB, error) {
> ```

Not for this PR, but we should consider the ergonomics we want. Right now, it looks like there may end up be a bunch of functions. But perhaps we want to be able to have `func (wkt WKT) ToWKB() (WKB, error)`. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.cc, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3w5NJK7qFEo6AXpiOO:-M3wAZZM5AqyReIqQIkS:b4kawnz) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb4039b9d9515c727f28954900a66ebf38f5064c/pkg/geo/geos/geos_unix.cc#L60)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

didn't realise we don't use this-> by default. is there any vague style guide we have for cpp at CRDB?
</blockquote></details>

We don't necessarily follow it, but this is the guide I grew up on: https://google.github.io/styleguide/cppguide.html


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46867)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan, @petermattis, and @sumeerbhola)

___
*[pkg/geo/geos/geos_unix.h, line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3w9_zb9pHxBDrVEPTM:-M3wFoNL5i_v7vPn0xFw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a3054d68b2a4f7109627c744f23b2fa45b15bc/pkg/geo/geos/geos_unix.h#L42)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

In `libroach` the difference between `DBSlice` and `DBString` is ownership. A returned `DBSlice` implies the underlying memory does not need to be freed. A returned `DBString` must have `DBString::data` freed by the caller of the function. I think it would be useful to maintain that distinction here. So `s/CR_GEOS_Slice/CR_GEOS_String/g`.
</blockquote></details>

Done.
___
*[pkg/geo/geos/geos_unix.go, line 115 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46867#-M3wAEEJA843fV-D3qnN:-M3wFtMD4LhR9IdcjeiW:bf5wjf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/82a3054d68b2a4f7109627c744f23b2fa45b15bc/pkg/geo/geos/geos_unix.go#L115)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Not for this PR, but we should consider the ergonomics we want. Right now, it looks like there may end up be a bunch of functions. But perhaps we want to be able to have `func (wkt WKT) ToWKB() (WKB, error)`. 
</blockquote></details>

most of the GEOS functionality will be in geomfn/ package, and in future this stuff may move there.
it's by itself at the moment because there's probably going to be a circular dependecy on geomfn and geo if we do it that way. WKTToWKB should also go away at one point since we only do this because we don't have  a WKT parser.


<!-- Sent from Reviewable.io -->

otan commented:
cool, thanks for the thorough review and help with the c-go stuff :) 

bors r=petermattis

otan commented:
bors r=petermattis

thanks bors for silently cancelling my PR :<. RIP GitHub just now.

```

{:timeout,  {GenServer, :call,   [     BorsNG.GitHub,     {:post_commit_status,      {{:installation, 117629}, 16563587},      {"01eb018092df6fec37490fbec3c1a06be16d3d13", :error,       "Canceled",       "https://bors.crdb.io/repositories/56/log#batch-9186"}},     10000   ]}}
--



```
